Mathlib/Tactic/RelCongr/Core.lean

```diff
@@ -0,0 +1,479 @@
+  match template with
+  | none =>
+    -- A. If there is no template, try to resolve the goal by the provided `assumption` tactic, and
+    -- continue on if this fails.
+    try assumption g; return (names, #[])
+    catch _ => pure ()
+  | some tpl =>
+    -- B. If there is a template:
+    -- (i) if the template is `?_` (or `?_ x1 x2`, created by entering binders)
+    -- then try to resolve the goal by the provided `assumption` tactic;
+    -- if this fails, stop and report the existing goal.
+    if let .mvar mvarId := tpl.getAppFn then
+      if let .syntheticOpaque ← mvarId.getKind then
+        try assumption g; return (names, #[])
+        catch _ => return (names, #[g])
+    -- (ii) if the template is *not* `?_` then continue on.
+  -- Check that the goal is of the form `rel (lhsHead _ ... _) (rhsHead _ ... _)`
+  let .app (.app rel lhs) rhs ← withReducible g.getType'
+    | throwError "rel_congr failed, not a relation"
+  let some relName := rel.getAppFn.constName?
+    | throwError "rel_congr failed, relation head {rel} is not a constant"
+  let (some lhsHead, lhsArgs) := lhs.withApp fun e a => (e.constName?, a)
+    | if template.isNone then return (names, #[g])
+      throwError "rel_congr failed, {lhs} is not a constant"
+  let (some rhsHead, rhsArgs) := rhs.withApp fun e a => (e.constName?, a)
+    | if template.isNone then return (names, #[g])
+      throwError "rel_congr failed, {rhs} is not a constant"
+  -- B. If there is a template, check that it is of the form `tplHead _ ... _` and that
+  -- `tplHead = lhsHead = rhsHead`
+  let tplArgs ← if let some tpl := template then
+    let (some tplHead, tplArgs) := tpl.withApp fun e a => (e.constName?, a)
+      | throwError "rel_congr failed, {tpl} is not a constant"
+    unless tplHead == lhsHead && tplArgs.size == rhsArgs.size do
+      throwError "expected {tplHead}, got {lhsHead}\n{lhs}"
+    unless tplHead == rhsHead && tplArgs.size == rhsArgs.size do
+      throwError "expected {tplHead}, got {rhsHead}\n{rhs}"
+    -- and also build a array of `Expr` corresponding to the arguments `_ ... _` to `tplHead` in the
+    -- template (these will be used in recursive calls later), and a array of booleans according to
+    -- which of these contain `?_`
+    tplArgs.mapM fun tpl => do
+      let mctx ← getMCtx
+      let hasMVar := tpl.findMVar? fun mvarId =>
+        if let some mdecl := mctx.findDecl? mvarId then
+          mdecl.kind matches .syntheticOpaque
+        else
+          false
+      pure (some tpl, hasMVar.isSome)
+  -- A. If there is no template, check that `lhs = rhs`
+  else
+    unless lhsHead == rhsHead && lhsArgs.size == rhsArgs.size do
+      -- (if not, stop and report the existing goal)
+      return (names, #[g])
+    -- and also build a array of booleans according to which arguments `_ ... _` to the head
+    -- function differ between the LHS and RHS
+    (lhsArgs.zip rhsArgs).mapM fun (lhsArg, rhsArg) =>
+      return (none, !(← isDefEq lhsArg rhsArg))
```

Comment:
We weren't sure about this one. Since this is a congr-like tactic I think we want to be able to prove subgoals where the two sides are semireducible defeq, but maybe it's too early to use it here and we should try this only on unsolved main goals.

Reply:
Should it at least be `withReducibleAndInstances`?  I'm not sure about that vs `isDefEq`, but I assume you want to identify defeq instance arguments?

Reply:
Oh, `withReducibleAndInstances` is better indeed.
I believe that in Lean 3 checking `isDefEq` on some completely different objects can be very expensive with semireducible transparancy, since it may involve unfolding *many* definitions. 
I think that the (nested) `isDefEq` checks in `relAssumption` should probably also be done with `withReducibleAndInstances`.

Reply:
Done!